File: CPP/ByRound/2034/2034_G1_Simurgh_s_Watch_Easy_Version.cpp
// Problem: CF 2034 G1 - Simurgh's Watch (Easy Version)
// https://codeforces.com/contest/2034/problem/G1

/*
Algorithm: Interval Coloring with Minimum Colors
Approach:
1. First, check if all intervals are non-overlapping - if so, only 1 color is needed
2. Otherwise, use sweep line algorithm combined with greedy coloring:
   - Create events for interval starts and ends
   - Sweep through events maintaining count of active intervals
   - Assign colors greedily to ensure unique color at each point
3. For overlapping intervals, use a 3-color assignment strategy:
   - If intervals fully overlap, try to use 2 colors
   - Otherwise, use 3 colors to ensure valid assignment

Time Complexity: O(n log n) due to sorting
Space Complexity: O(n) for storing events and results
*/
#include <stdlib.h>
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

struct ev;

using namespace std;
typedef long long ll;
typedef long double ld;
typedef pair<int, int> ii;
typedef pair<ll, ll> pll;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef vector<ll> vll;
typedef vector<pll> vpll;
#define PB push_back
#define PF push_front
#define PPB pop_back
#define PPF pop_front
#define X first
#define Y second
#define MP make_pair
#define all(x) (x).begin(), (x).end()
const int mod = 1e9 + 7;
const int inf = 1e9 + 7;
const ll INF = 1e18 + 7;
const int logo = 20;
const int MAXN = 1e6 + 7;
const int off = 1 << logo;
const int trsz = off << 1;
const int dx[] = {1, -1, 0, 0};
const int dy[] = {0, 0, -1, 1};
struct ev {
  int x, y, i;
};
bool cmp(ev &a, ev &b) {
  if (a.x != b.x)
    return a.x < b.x;
  if (a.y != b.y)
    return a.y < b.y;
  return a.i < b.i;
}
vector<ev> a, e;
int ans[MAXN];
void solve() {
  int n;
  cin >> n;
  a.clear(), e.clear();
  for (int i = 1; i <= n; i++) {
    int l, r;
    cin >> l >> r;
    a.PB({l, r, i});
  }
  sort(all(a), cmp);
  bool dobar = 1;
  for (int i = 1; i < n; i++) {
    if (a[i].x <= a[i - 1].y)
      dobar = 0;
  }
  if (dobar) {
    cout << 1 << "\n";
    for (int i = 0; i < n; i++)
      cout << 1 << " ";
    cout << "\n";
    return;
  }
  for (auto &x : a) {
    e.PB({x.x, -x.y, -x.i});
    e.PB({x.y, -x.y, x.i});
  }
  sort(all(e), cmp);
  int c[3] = {0, 0, 0}, to[3] = {0, 0, 0};
  dobar = 1;
  int zadnji = -1;
  for (auto [p, _, i] : e) {
    if (p != zadnji and c[1] != 1 and c[2] != 1 and c[1] + c[2] > 0) {
      dobar = 0;
      break;
    }
    zadnji = p;
    if (i < 0) {
      int ja = 1;
      if (c[1] == 0 or c[1] > 1)
        ja = 1;
      else if (c[2] == 0 or c[2] > 1)
        ja = 2;
      else if (to[1] < to[2])
        ja = 1;
      else
        ja = 2;
      c[ja]++;
      to[ja] = max(to[ja], abs(_));
      ans[-i] = ja;
    } else
      c[ans[i]]--;
  }
  if (dobar) {
    cout << 2 << "\n";
    for (int i = 1; i <= n; i++)
      cout << ans[i] << " ";
    cout << "\n";
    return;
  }
  for (int i = 1; i <= n; i++)
    ans[i] = 0;
  zadnji = -1;
  int col = 1;
  for (int i = 0; i < n;) {
    if (zadnji < a[i].x)
      zadnji = a[i].x;
    int tko = i;
    while (i < n and a[i].x <= zadnji) {
      if (a[i].y > a[tko].y)
        tko = i;
      i++;
    }
    ans[a[tko].i] = col;
    zadnji = a[tko].y;
    col ^= 3;
  }
  cout << "3\n";
  for (int i = 1; i <= n; i++)
    cout << ans[i] + 1 << " ";
  cout << "\n";
}
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  int tt = 1;
  cin >> tt;
  while (tt--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/G/2034_G1_Simurgh_s_Watch_Easy_Version.py
# Problem: CF 2034 G1 - Simurgh's Watch (Easy Version)
# https://codeforces.com/contest/2034/problem/G1

### Code Purpose:
#This program solves the problem of coloring intervals (warrior watch segments) such that for every point in time overlapping with at least one interval, there exists at least one color that appears on exactly one interval at that point. 

### Algorithms/Techniques:
#1. **Sweep Line Algorithm** - to process intervals based on their start and end boundaries.
#2. **Greedy Color Assignment** - assigning colors based on availability and conflict resolution during sweep.
#3. **Fallback to Three Colors** - when two colors cannot resolve overlapping conflicts, a third color is utilized.

### Time and Space Complexity:
#- **Time Complexity**: $ O(n \log n) $, dominated by sorting intervals and events.
#- **Space Complexity**: $ O(n) $, for storing intervals, events, and result arrays.

import sys


def solve():
    # Read all input data at once for efficiency
    it = iter(map(int, sys.stdin.buffer.read().split()))
    t = next(it)
    o = []
    for _ in range(t):
        n = next(it)
        x = [0] * n
        y = [0] * n
        for i in range(n):
            x[i] = next(it)
            y[i] = next(it)
        # Sort intervals by start time, then end time, then index
        ord = sorted(range(n), key=lambda i: (x[i], y[i], i))
        # Check if intervals are non-overlapping
        iv = any(x[ord[i]] <= y[ord[i - 1]] for i in range(1, n))
        if not iv:
            # All intervals are non-overlapping, can use 1 color
            o.append("1")
            o.append(" ".join(["1"] * n))
            continue
        # Create event list for sweep line algorithm
        ev = []
        for i in range(n):
            # Start event: (start_time, -end_time, ~index) 
            # Negative end_time to sort properly, ~index for unique identification
            ev.append((x[i], -y[i], ~i))
            # End event: (end_time, -end_time, index)
            ev.append((y[i], -y[i], i))
        ev.sort()
        z = [0] * n  # Color assignment array
        c = [0, 0, 0]  # Counters for colors (1 and 2)
        f = [0, 0, 0]  # Furthest end_time for each color
        bad = False
        ls = None  # Last event time
        for w, _, q in ev:
            # If time changed, check if we're in a valid state for current coloring
            if ls is not None and w != ls:
                if c[1] != 1 and c[2] != 1 and (c[1] + c[2] > 0):
                    bad = True
                    break
            ls = w
            if q < 0:
                # Start event for interval i (~q is index in original array)
                i = ~q
                # Choose color, preferring color 1 or 2 if only one has 0 or >1 count
                if c[1] == 0 or c[1] > 1:
                    p = 1
                elif c[2] == 0 or c[2] > 1:
                    p = 2
                elif f[1] < f[2]:
                    p = 1
                else:
                    p = 2
                c[p] += 1
                if y[i] > f[p]:
                    f[p] = y[i]
                z[i] = p
            else:
                # End event for interval q (index in original array)
                c[z[q]] -= 1
        if not bad:
            # Successfully assigned 2 colors
            o.append("2")
            o.append(" ".join(map(str, z)))
            continue
        # If two colors is not sufficient, use three colors
        z = [0] * n
        i = 0
        ls = -(10**20)  # Last end_time processed
        nw = 1  # Current color (1 or 2, toggled with ^3)
        while i < n:
            if ls < x[ord[i]]:
                ls = x[ord[i]]
            bs = i
            # Find all overlapping intervals that have the same start or overlap
            while i < n and x[ord[i]] <= ls:
                if y[ord[i]] > y[ord[bs]]:
                    bs = i
                i += 1
            idx = ord[bs]
            z[idx] = nw
            ls = y[idx]
            nw ^= 3  # Toggle between 1 and 2
        o.append("3")
        o.append(" ".join(str(c + 1) for c in z))
    sys.stdout.write("\n".join(o))


if __name__ == "__main__":
    solve()

# https://github.com/VaHiX/CodeForces/